Feature/key provisioning

sukhmanm opened the pull request:
Key Provisioning Demo (Internally Generated Keys)

Description
-----------
Demo to demonstrate how users can generate private keys on device instead of importing into header file. Option to run old demo using #define variable.

Demo located in demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c
Script to create CA certificate in tools/create_certs/generate_CA.sh
Script to create device certificate in tools/create_certs/device_cert.sh


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

alexa-noxon commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+#define pkcs11testLABEL_DEVICE_PUBLIC_KEY_FOR_TLS       pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS
+
+#define SHA256_DIGEST_SIZE                              32
+#define ECDSA_SIGNATURE_SIZE                            64
```
For these defines, we can include iot_pkcs11.h and use pkcs11SHA256_DIGEST_LENGTH, pkcs11ECDSA_P256_SIGNATURE_LENGTH, and  pkcs11RSA_2048_SIGNATURE_LENGTH respectively.

This is making me realize that I was creating some un-needed defines in the test code which I should remove :)

sukhmanm replied:
Modified to use iot_pkcs11.h defines instead and removed a few I didn't end up using. Thanks!

alexa-noxon commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+#define SHA256_DIGEST_SIZE                              32
+#define ECDSA_SIGNATURE_SIZE                            64
+#define RSA_SIGNATURE_SIZE                              256
+#define MBEDTLS_X509_CSR_H
```
Are these also defined in mbedtls/config.h?  I would check that they are, and then deleting them from this file

sukhmanm replied:
Yes they are -- removed the redundancy

alexa-noxon commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+ *  Prints error message upon failure, success message upon success
+ *
+ */
+static void vProvisionDevice( void )
```
I would recommend returning a non-void value here.

It has been challenging debugging our void returns for other provisioning functions :)

sukhmanm replied:
Changed to an int return value :)

alexa-noxon commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+ *  Prints error message upon failure, success message upon success
+ *
+ */
+static void vProvisionDevice( void )
```
Consider renaming this function xProvisionDeviceForJITP so that it is not confused with the devmodekeyprovisioning function xProvisionDevice.

Same below for vDevModeDeviceKeyProvisioning ... let's highlight the difference between this function and the one that already exists with their functions
This makes me realize that perhaps I should rename vDevModeKeyProvisioning and  xProvisionDevice as well :)

sukhmanm replied:
renamed!

alexa-noxon commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to destroy credentials before Generating Key Pair.\r\n" ) );
```
For this (and other configPRINTFs in this file) consider printing the error code returned to help with debugging using logs.

For example, configPRINTF( ( "ERROR: %d - Failed to destroy credentials before Generating Key Pair.\r\n", xResult ) );

sukhmanm replied:
Modified all my config outputs to include error codes! Thanks.

alexa-noxon commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+            /* Initializing CSR Context */
+            mbedtls_x509write_csr my_csr;
+            mbedtls_x509write_csr_init( &my_csr );
+            ret = mbedtls_x509write_csr_set_subject_name( &my_csr, "CN=ThingName" );
```
I'd suggest highlighting that this thing name is configurable (maybe we pull it from clientcredential, or from a define)

Also, from the template, is the thing name being pulled from the common name or the certificate ID?  For this example, I think common name is the way to go like you have done here :)

sukhmanm replied:
Added a comment regarding thing name configuration.

Yes, I originally had the thing name pulled from the cert ID in the template -- I just modified it to pull from the common name instead :)

alexa-noxon commented on tools/create_certs/generate_CA_backup.sh:
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
```
Do you still need anything from this backup file?  If not, let's remove it so future us don't get confused :)

sukhmanm replied:
deleted!

alexa-noxon commented on tools/create_certs/provisioning-template.json:
```diff
@@ -0,0 +1,4 @@
+{
+ "templateBody":"{ \"Parameters\" : { \"AWS::IoT::Certificate::Country\" : { \"Type\" : \"String\" }, \"AWS::IoT::Certificate::Id\" : { \"Type\" : \"String\" } }, \"Resources\" : { \"thing\" : { \"Type\" : \"AWS::IoT::Thing\", \"Properties\" : { \"ThingName\" : {\"Ref\" : \"AWS::IoT::Certificate::Id\"}, \"AttributePayload\" : { \"version\" : \"v1\", \"country\" : {\"Ref\" : \"AWS::IoT::Certificat ...
```
Not 100% sure I'm reading this template correctly, but if it's using Certificate ID as the thing name, pet's use common name instead so that it was controlled by the cert creator in the CSR, and no one has to reverse engineer how certificate ID is generated.

sukhmanm replied:
Changed to using common name for thing name

alexa-noxon commented on tools/create_certs/rootCA.srl:
```diff
@@ -0,0 +1 @@
+87F5115738EE8D42
```
How did this serial number get chosen?  Do we want all certs to have it?

sukhmanm replied:
Good point -- I think it is unnecessary for the demo. Thanks!

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+/*-----------------------------------------------------------*/
+
+/* @brief Alternate Signing Function to be Passed into PK Context Header */
+static int prvPrivateKeySigningCallback( void * pvContext,
```
Is this copied from somewhere else? That's understandable from the perspective of expedience for this demo. But it would be helpful to have a comment about where it came from, in case there's subsequent interest in refactoring.

sukhmanm replied:
Yes, it comes from iot_tls.c. Just added a comment to clarify.

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+        xResult = CKR_ARGUMENTS_BAD;
+    }
+
+    xMech.mechanism = CKM_ECDSA;
```
Is this routine ECDSA-only? The defines at the top of this file led me to believe that the demo also supports RSA.

sukhmanm replied:
Yes, it's ECDSA only. Removed all unnecessary RSA defines. Thanks.

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+        {
+            CredentialsProvisioned_t xCurrentCredentials = eStateUnknown;
+
+            /* Note that mbedTLS does not permit a signature on all 0's. */
```
Technically, I think the issue is what happens when performing the resulting exponentiation in ECDSA - so the problem is the math, not the library.

sukhmanm replied:
I copied this comment from a test file, but removed it to avoid any confusion. Thanks.

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+                configPRINTF( ( "Failed to ECDSA Sign.\r\n" ) );
+            }
+
+            /* Verify the signature with mbedTLS */
```
I think that round-tripping the new key between a PKCS #11 signature and an mbedTLS verification would in fact be a great addition here. It doesn't look like that's happening, though, at least not in this immediate few lines of code.

sukhmanm replied:
Added!

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+            const mbedtls_pk_info_t * header = mbedtls_pk_info_from_type( type_key );
+
+            /* Creating copy of header to pass in custom sign function (original header is immutable) */
+            mbedtls_pk_info_t * header_copy = pvPortMalloc( sizeof( mbedtls_pk_info_t ) );
```
Need to check for resource exhaustion.

sukhmanm replied:
Added Null checks for malloc calls. Thanks for the pointer!

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+            }
+
+            /* Output Buffer */
+            unsigned char * final_csr = pvPortMalloc( 2000 );
```
Check for resource exhaustion.

sukhmanm replied:
Added null check. Thanks again :)

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+            mbedtls_ecdsa_free( &xEcdsaContext );
+            /*mbedtls_pk_free( &pk_cont ); */
+            mbedtls_x509write_csr_free( &my_csr );
+            vPortFree( final_csr );
```
Is header_copy being leaked?

sukhmanm replied:
Yes, I missed that. Added a line to free it.

sukhmanm replied:
^ Still working on this. It appears that not all contents of header_copy are being freed correctly, which is causing a memory leak. Will continue to look into it. Thanks.

dcgaws replied:
I don't see the call to free header_copy. Is that in a separate commit that you haven't pushed yet?

jainamt replied:
Yes, this has not yet been addressed. Sukhmani was working on fixing it yesterday but still needs to continue on it. There will be another patch coming up for this.

dcgaws commented on libraries/3rdparty/mbedtls/include/mbedtls/config.h:
```diff
@@ -1075,7 +1075,7 @@
  * Enable functions that use the filesystem.
  */
-//#define MBEDTLS_FS_IO
+#define MBEDTLS_FS_IO
```
We would also want to avoid changing this global config file anyway. Our approach with that is admittedly limiting, but I suspect that this demo could find a way to override the settings it needs.

sukhmanm replied:
Yes, we don't need that define anymore -- just commented it back out. 

@Dan - I needed to uncomment some of the defines in the config file for CSR-related mbedTLS calls. I'm not sure if there is an effective workaround, but if so I'll look into implementing that.

dcgaws commented on tools/create_certs/root.cert:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN CERTIFICATE-----
```
Which root is this?

sukhmanm replied:
It was used when performing JITP manually, but is unnecessary now. Just deleted :)

dcgaws replied:
It looks like this root certificate file is still being added as part of your PR. Do you still have a pending commit that removes it?

sukhmanm replied:
That's my bad. I had incorrectly removed the file from remote and not local, and then force pushed my branch. Just updated. Thanks for pointing that out.

sukhmanm commented on demos/common/mqtt/provisioning_scripts/generate_CA.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
```
Deleted!

sukhmanm commented on tools/create_certs/provisioning-template.json:
```diff
@@ -0,0 +1,4 @@
+{
+ "templateBody":"{ \"Parameters\" : { \"AWS::IoT::Certificate::Country\" : { \"Type\" : \"String\" }, \"AWS::IoT::Certificate::Id\" : { \"Type\" : \"String\" } }, \"Resources\" : { \"thing\" : { \"Type\" : \"AWS::IoT::Thing\", \"Properties\" : { \"ThingName\" : {\"Ref\" : \"AWS::IoT::Certificate::Id\"}, \"AttributePayload\" : { \"version\" : \"v1\", \"country\" : {\"Ref\" : \"AWS::IoT::Certificat ...
```
Yes -- just added those configurations to the template!

sukhmanm commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+    eRsaTest,             /* Provisioned using the RSA test credentials located in this file. */
+    eEllipticCurveTest,   /* Provisioned using EC test credentials located in this file. */
+    eClientCredential,    /* Provisioned using the credentials in aws_clientcredential_keys. */
+    eGeneratedEc,         /* Provisioned using elliptic curve generated on device.  Private key unknown.  No corresponding certificate. */
```
Deleted this structure. Was used for testing purposes but we don't need it anymore :)

sukhmanm commented on demos/common/mqtt/provisioning_scripts/generate_CA.sh:
```diff
@@ -0,0 +1,44 @@
+commonname="."
+email="."
+
+password=""
```
Yes, the script seems to work even when protected.

sukhmanm commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+ *  \return 0 on success
+ *
+ */
+static int prvRNG( void * pkcs_session,
```
I pass this function into mbedtls_x509write_csr_pem, which requires a RNG of a certain type/format that configRAND32 unfortunately doesn't seem to match.

alexa-noxon commented on tools/ca_certificate/provisioning-template.json:
```diff
@@ -1,4 +0,0 @@
-{
- "templateBody":"{ \"Parameters\" : { \"AWS::IoT::Certificate::Country\" : { \"Type\" : \"String\" }, \"AWS::IoT::Certificate::Id\" : { \"Type\" : \"String\" } }, \"Resources\" : { \"thing\" : { \"Type\" : \"AWS::IoT::Thing\", \"Properties\" : { \"ThingName\" : {\"Ref\" : \"AWS::IoT::Certificate::Id\"}, \"AttributePayload\" : { \"version\" : \"v1\", \"country\" : {\"Ref\" : \"AWS::IoT::Certificat ...
- "roleArn":"arn:aws:iam::152386550361:role/JITPRole"
```
I think that we will need a provisioning template of some sort checked in to git, but in future work we may need to script generating it using user input from JITPRole and account number.

sukhmanm replied:
There is a provisioning template in tools/create_certs. I realized tools/ca_certificate was outdated/unnecessary so I removed the folder :)

sukhmanm commented on demos/common/mqtt/provisioning_scripts/generate_CA.sh:
```diff
@@ -0,0 +1,44 @@
+echo "Registering CA Certificate with AWS IoT"
+
+registration_code_str=$(aws iot get-registration-code)  # registraton code / common name
+registration_code="${registration_code_str:27:64}" # CHECK THIS: REGISTRATION CODE LENGTH MAY NOT BE THE SAME ALWAYS
```
Yes, I updated it to parse from the object name instead. Thanks!

sukhmanm commented on demos/common/mqtt/provisioning_scripts/generate_CA.sh:
```diff
@@ -0,0 +1,44 @@
+echo "Creating CA Certificate"
+
+# Pass-in Variables
+country=US
```
Added a comment!

sukhmanm commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,613 @@
+
+
+#define pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS    "Device Priv TLS Key"
+#define pkcs11testLABEL_DEVICE_PRIVATE_KEY_FOR_TLS      pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS
```
Deleted the test defines!

sukhmanm commented on demos/common/mqtt/provisioning_scripts/generate_CA.sh:
```diff
@@ -0,0 +1,44 @@
+
+# CA Certificate
+echo "Generating CA Certificate"
+openssl req -x509 -config rootCA_openssl.conf -new -nodes -key rootCA.key -sha256 -extensions v3_ca -days 1024 -out rootCA.pem -passin pass:$password \
```
I considered doing this, but for debugging simplicity decided to contain everything in the script. Going forward, however, I can modify the approach so the user doesn't have to edit the script. Thanks!

sukhmanm commented:
Apart from the memory leak issue and the two tasks that I've deferred (see unresolved above - I've added these tasks to my current sprint at https://sim.amazon.com/issues/TS-8256), I believe I've addressed all comments. Feel free to take a look at my updates when you get a chance. Thank you -- I appreciate it. :)

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,675 @@
+            }
+
+            /* Output Buffer */
+            unsigned char * final_csr = pvPortMalloc( 2000 );
```
A nicer way to handle this would be to have an inline #define for the length of the buffer. Then declare len_buf and initialize it to the value of the #define. Then call malloc, passing in len_buf. In my experience, that ordering is less likely to lead to mistakes down the road.

sukhmanm replied:
Thank you. I added the #define for the buffer size and used that instead.

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,675 @@
+            mbedtls_x509write_csr_free( &my_csr );
+            vPortFree( final_csr );
+
+            xHeapAfter = xPortGetFreeHeapSize();
```
Good idea here. How many bytes are you leaking?

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,675 @@
+ *  \return Void function, console outputs upon success/failure
+ *
+ */
+static void xDeviceProvisioningForJITP( void )
```
I think that there had been a comment at some point about changing the signature of the various "DevModeProvisioning" routines to return status codes. I recommend that change, since it makes it easier for the application writer to isolate problems (say, if they start using a secure element instead).

sukhmanm replied:
I changed all provisioning functions to return integer error code values. Thank you!

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c:
```diff
@@ -0,0 +1,675 @@
+/* Standard includes. */
```
I don't see the modified solution/project files as part of this PR. Can you please include them?

jainamt replied:
I had given direction to Sukhmani to add that after her demo today. We acknowledge those files need to be added, but with the timing of her demo and presentation this afternoon, she will handle that change later today.

sukhmanm replied:
All project files should be updated now :) Thanks!

sukhmanm commented:
Closing this PR. Thank you for all your help! We've made another branch "sukhmanm/jitpdemo" that contains the following core files to be picked off of for future use:

tools/create_certs/deviceandCAcerts.sh
tools/create_certs/.gitignore
libraries/3rdparty/mbedtls/include/mbedtls/config.h
demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c
demos/include/iot_demo_runner.h

Thanks again! :)
